pkg/controllers/dnspolicy/dnspolicy_healthchecks.go

```diff
@@ -22,91 +26,108 @@ type healthChecksConfig struct {
 	Protocol         *dns.HealthCheckProtocol
 }
 
-func (r *DNSPolicyReconciler) ReconcileHealthChecks(ctx context.Context, log logr.Logger, policy *v1alpha1.DNSPolicy) error {
-	// Get the associated DNSRecord
-	dnsRecords, err := r.getDNSRecords(ctx, policy)
-	if err != nil {
-		return err
+func (r *DNSPolicyReconciler) reconcileHealthChecks(ctx context.Context, dnsPolicy *v1alpha1.DNSPolicy, gwDiffObj *reconcilers.GatewayDiff) error {
+	log := crlog.FromContext(ctx)
+
+	// Delete Health checks for each gateway no longer referred by this policy
+	for _, gw := range gwDiffObj.GatewaysWithInvalidPolicyRef {
+		log.V(1).Info("reconcileHealthChecks: gateway with invalid policy ref", "key", gw.Key())
+		_, err := r.reconcileGatewayHealthChecks(ctx, gw.Gateway, nil)
+		if err != nil {
+			return err
+		}
 	}
-	if len(dnsRecords) == 0 {
-		log.Info("No DNSRecord found matching policy", "policy", policy)
-		return nil
+
+	healthCheckConfig := getHealthChecksConfig(dnsPolicy)
+	allResults := []dns.HealthCheckResult{}
+
+	// Reconcile Health checks for each gateway directly referred by the policy (existing and new)
+	for _, gw := range append(gwDiffObj.GatewaysWithValidPolicyRef, gwDiffObj.GatewaysMissingPolicyRef...) {
```

Comment:
so again not sure I am following this, a policy can only target one resource, so if the target changes is not the case that we should delete the associated health checks? I am probably misunderstanding what is happening here